File: ScheduleSync.Core/Parsers/StratusCsvUpdateSource.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using ScheduleSync.Core.Interfaces;
using ScheduleSync.Core.Mapping;
using ScheduleSync.Core.Models;

namespace ScheduleSync.Core.Parsers
{
    /// <summary>
    /// Parses task updates from a STRATUS Packages Dashboard CSV export.
    /// Maps STRATUS-specific column names to <see cref="TaskUpdate"/> fields
    /// and resolves fabrication status to percent-complete via <see cref="StatusPercentMap"/>.
    /// </summary>
    /// <remarks>
    /// Expected columns (case-insensitive):
    ///   Number, Name, Project Number Override (fallback: Project Number),
    ///   Prefab Build Start Date, Prefab Build Finish Date,
    ///   Work Days (Reference), Required, Status, Notes,
    ///   Location, Description, Cost Code Category, Category Type,
    ///   STRATUS.Package.Id
    /// </remarks>
    public class StratusCsvUpdateSource : IUpdateSource
    {
        /// <summary>Minutes per work day (8-hour day).</summary>
        public const int MinutesPerWorkDay = 480;

        public ParseResult Parse(string content)
        {
            var result = new ParseResult();
            if (string.IsNullOrWhiteSpace(content))
            {
                result.Errors.Add(new ParseError(null, null, "CSV content is empty."));
                return result;
            }

            var lines = CsvParserHelper.ReadLines(content);
            if (lines.Count < 2)
            {
                result.Errors.Add(new ParseError(null, null, "CSV must contain a header row and at least one data row."));
                return result;
            }

            var headers = CsvParserHelper.ParseCsvLine(lines[0])
                .Select(h => h.Trim().ToLowerInvariant())
                .ToList();

            for (int i = 1; i < lines.Count; i++)
            {
                if (string.IsNullOrWhiteSpace(lines[i])) continue;

                int rowNumber = i + 1;
                var fields = CsvParserHelper.ParseCsvLine(lines[i]);

                if (fields.Count != headers.Count)
                {
                    result.Errors.Add(new ParseError(rowNumber, null,
                        $"Row has {fields.Count} fields but header has {headers.Count}."));
                    continue;
                }

                var lookup = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
                for (int j = 0; j < headers.Count; j++)
                {
                    lookup[headers[j]] = fields[j].Trim();
                }

                var update = new TaskUpdate();
                bool hasError = false;

                // ── Composite external key: ProjectNumber-PackageNumber ──
                string projectNumber = GetFirstNonEmpty(lookup, "project number override", "project number");
                string packageNumber = GetValue(lookup, "number");

                if (string.IsNullOrEmpty(packageNumber))
                {
                    result.Errors.Add(new ParseError(rowNumber, "Number",
                        "Package Number is required for external key."));
                    hasError = true;
                }
                else
                {
                    update.ExternalKey = string.IsNullOrEmpty(projectNumber)
                        ? packageNumber
                        : $"{projectNumber}-{packageNumber}";
                }

                // ── Name ──
                update.Name = GetValue(lookup, "name");

                // ── Prefab Build Start Date → NewStart ──
                if (TryGetDate(lookup, "prefab build start date", rowNumber, result, out var startDate))
                    update.NewStart = startDate;

                // ── Prefab Build Finish Date → NewFinish ──
                if (TryGetDate(lookup, "prefab build finish date", rowNumber, result, out var finishDate))
                    update.NewFinish = finishDate;

                // ── Work Days (Reference) → NewDurationMinutes ──
                string workDaysStr = GetValue(lookup, "work days (reference)");
                if (!string.IsNullOrEmpty(workDaysStr))
                {
                    if (double.TryParse(workDaysStr, NumberStyles.Any, CultureInfo.InvariantCulture, out var workDays))
                    {
                        update.NewDurationMinutes = workDays * MinutesPerWorkDay;
                    }
                    else
                    {
                        result.Errors.Add(new ParseError(rowNumber, "Work Days (Reference)",
                            $"Invalid number: '{workDaysStr}'."));
                        hasError = true;
                    }
                }

                // ── Required → NewDeadline ──
                if (TryGetDate(lookup, "required", rowNumber, result, out var deadline))
                    update.NewDeadline = deadline;

                // ── Status → NewPercentComplete ──
                string status = GetValue(lookup, "status");
                if (!string.IsNullOrEmpty(status))
                {
                    var pct = StatusPercentMap.Resolve(status);
                    if (pct.HasValue)
                    {
                        update.NewPercentComplete = pct.Value;
                    }
                    else
                    {
                        result.Errors.Add(new ParseError(rowNumber, "Status",
                            $"Unrecognized status: '{status}'. Percent complete will not be set."));
                        // Non-fatal: don't set hasError, just warn via parse errors
                    }
                }

                // ── Notes + Description → NotesAppend ──
                string notes = GetValue(lookup, "notes");
                string description = GetValue(lookup, "description");
                var notesParts = new List<string>();
                if (!string.IsNullOrEmpty(description)) notesParts.Add(description);
                if (!string.IsNullOrEmpty(notes)) notesParts.Add(notes);
                if (notesParts.Count > 0)
                    update.NotesAppend = string.Join("\n", notesParts);

                // ── Metadata for hierarchy and categorization ──
                if (!string.IsNullOrEmpty(projectNumber))
                    update.Metadata["ProjectNumber"] = projectNumber;

                string location = GetValue(lookup, "location");
                if (!string.IsNullOrEmpty(location))
                    update.Metadata["Location"] = location;

                string categoryType = GetValue(lookup, "category type");
                if (!string.IsNullOrEmpty(categoryType))
                    update.Metadata["CategoryType"] = categoryType;

                string costCodeCategory = GetValue(lookup, "cost code category");
                if (!string.IsNullOrEmpty(costCodeCategory))
                    update.Metadata["CostCodeCategory"] = costCodeCategory;

                string costCodeNumber = GetValue(lookup, "cost code number");
                if (!string.IsNullOrEmpty(costCodeNumber))
                    update.Metadata["CostCodeNumber"] = costCodeNumber;

                string stratusId = GetValue(lookup, "stratus.package.id");
                if (string.IsNullOrEmpty(stratusId))
                    stratusId = GetValue(lookup, "id");
                if (!string.IsNullOrEmpty(stratusId))
                    update.Metadata["StratusPackageId"] = stratusId;

                if (!hasError)
                    result.Updates.Add(update);
            }

            return result;
        }

        /// <summary>
        /// Detects whether a CSV header row belongs to a STRATUS Packages Dashboard export.
        /// </summary>
        public static bool IsStratusFormat(string headerLine)
        {
            if (string.IsNullOrWhiteSpace(headerLine))
                return false;

            var headers = CsvParserHelper.ParseCsvLine(headerLine)
                .Select(h => h.Trim().ToLowerInvariant())
                .ToList();

            // Check for characteristic STRATUS columns
            return headers.Contains("prefab build start date")
                || headers.Contains("stratus.package.id")
                || headers.Contains("cost code number");
        }

        #region Helpers

        private static string GetValue(Dictionary<string, string> lookup, string key)
        {
            return lookup.TryGetValue(key, out var val) ? (string.IsNullOrEmpty(val) ? null : val) : null;
        }

        private static string GetFirstNonEmpty(Dictionary<string, string> lookup, params string[] keys)
        {
            foreach (var key in keys)
            {
                var val = GetValue(lookup, key);
                if (!string.IsNullOrEmpty(val))
                    return val;
            }
            return null;
        }

        private static bool TryGetDate(
            Dictionary<string, string> lookup, string key, int rowNumber,
            ParseResult result, out DateTime date)
        {
            date = default;
            if (!lookup.TryGetValue(key, out var str) || string.IsNullOrEmpty(str?.Trim()))
                return false;

            str = str.Trim();
            if (DateTime.TryParse(str, CultureInfo.InvariantCulture, DateTimeStyles.None, out date))
                return true;

            // Try US date formats commonly used in STRATUS exports
            if (DateTime.TryParseExact(str,
                new[] { "M/d/yyyy", "MM/dd/yyyy", "M/d/yyyy h:mm tt", "MM/dd/yyyy hh:mm tt" },
                CultureInfo.InvariantCulture, DateTimeStyles.None, out date))
                return true;

            result.Errors.Add(new ParseError(rowNumber, key, $"Invalid date: '{str}'."));
            return false;
        }

        #endregion
    }
}

File: web/packages/engine/src/parsers/stratusCsvUpdateSource.ts
/**
 * STRATUS CSV update source — ported from C# StratusCsvUpdateSource.
 * Maps STRATUS Packages Dashboard columns to TaskUpdate fields.
 */

import type { IUpdateSource, ParseResult, TaskUpdate, ParseError } from '../types.js';
import { createParseResult } from '../types.js';
import { parseCsvLine, readLines } from './csvParserHelper.js';
import { resolveStatus } from './statusPercentMap.js';

const MINUTES_PER_WORK_DAY = 480;

function makeEmptyUpdate(): TaskUpdate {
  return {
    externalKey: null,
    uniqueId: null,
    name: null,
    newStart: null,
    newFinish: null,
    newDurationMinutes: null,
    newPercentComplete: null,
    newConstraintType: null,
    newConstraintDate: null,
    newDeadline: null,
    notesAppend: null,
    allowConstraintOverride: false,
    isNew: true, // Stratus tasks are new by default
    metadata: {},
  };
}

/**
 * Attempt to parse a date string using multiple formats.
 * Tries ISO first, then US date formats (M/d/yyyy etc.).
 */
function tryParseDate(value: string): Date | null {
  if (!value || value.trim() === '') return null;

  // Try native Date parsing first (handles ISO and many formats)
  const d = new Date(value);
  if (!isNaN(d.getTime())) return d;

  // Try US date formats: M/d/yyyy, MM/dd/yyyy, M-d-yyyy, MM-dd-yyyy
  const usPatterns = [
    /^(\d{1,2})[\/\-](\d{1,2})[\/\-](\d{4})$/,
    /^(\d{1,2})[\/\-](\d{1,2})[\/\-](\d{4})\s+(\d{1,2}):(\d{2})(?::(\d{2}))?$/,
  ];

  for (const pattern of usPatterns) {
    const match = value.match(pattern);
    if (match) {
      const month = parseInt(match[1], 10);
      const day = parseInt(match[2], 10);
      const year = parseInt(match[3], 10);
      const hour = match[4] ? parseInt(match[4], 10) : 0;
      const minute = match[5] ? parseInt(match[5], 10) : 0;
      const second = match[6] ? parseInt(match[6], 10) : 0;
      const result = new Date(Date.UTC(year, month - 1, day, hour, minute, second));
      if (!isNaN(result.getTime())) return result;
    }
  }

  return null;
}

/**
 * Detect whether a CSV header line is in STRATUS format.
 * Returns true if headers contain distinctive STRATUS column names.
 */
export function isStratusFormat(headerLine: string | null | undefined): boolean {
  if (!headerLine) return false;
  const lower = headerLine.toLowerCase();
  return (
    lower.includes('prefab build start date') ||
    lower.includes('stratus.package.id') ||
    lower.includes('cost code number')
  );
}

export class StratusCsvUpdateSource implements IUpdateSource {
  parse(content: string): ParseResult {
    const updates: TaskUpdate[] = [];
    const errors: ParseError[] = [];

    const lines = readLines(content);
    if (lines.length === 0) {
      errors.push({ rowNumber: null, fieldName: null, message: 'Content is empty.' });
      return createParseResult(updates, errors);
    }

    const headers = parseCsvLine(lines[0]).map((h) => h.toLowerCase().trim());

    // Build column index map
    const colIndex = new Map<string, number>();
    for (let i = 0; i < headers.length; i++) {
      colIndex.set(headers[i], i);
    }

    if (lines.length <= 1) {
      errors.push({ rowNumber: null, fieldName: null, message: 'No data rows found.' });
      return createParseResult(updates, errors);
    }

    for (let i = 1; i < lines.length; i++) {
      const line = lines[i];
      if (line.trim() === '') continue;

      const fields = parseCsvLine(line);
      const rowNum = i + 1;
      const update = makeEmptyUpdate();

      const getField = (name: string): string =>
        fields[colIndex.get(name) ?? -1]?.trim() ?? '';

      // Number (required)
      const number = getField('number');
      if (!number) {
        errors.push({
          rowNumber: rowNum,
          fieldName: 'Number',
          message: 'Missing required Number field.',
        });
        continue;
      }

      // Project Number: prefer "project number override", fallback "project number"
      let projectNumber = getField('project number override');
      if (!projectNumber) {
        projectNumber = getField('project number');
      }

      // Composite external key: "ProjectNumber-Number"
      update.externalKey = projectNumber ? `${projectNumber}-${number}` : number;

      // Name
      update.name = getField('name') || null;

      // Metadata
      if (projectNumber) update.metadata['ProjectNumber'] = projectNumber;
      const location = getField('location');
      if (location) update.metadata['Location'] = location;
      const categoryType = getField('category type');
      if (categoryType) update.metadata['CategoryType'] = categoryType;
      const costCodeCategory = getField('cost code category');
      if (costCodeCategory) update.metadata['CostCodeCategory'] = costCodeCategory;
      const costCodeNumber = getField('cost code number');
      if (costCodeNumber) update.metadata['CostCodeNumber'] = costCodeNumber;

      // STRATUS.Package.Id — try "stratus.package.id" first, fallback to "id"
      let packageId = getField('stratus.package.id');
      if (!packageId) packageId = getField('id');
      if (packageId) update.metadata['StratusPackageId'] = packageId;

      // Dates
      const startStr = getField('prefab build start date');
      if (startStr) {
        const d = tryParseDate(startStr);
        if (d) {
          update.newStart = d.toISOString();
        } else {
          errors.push({ rowNumber: rowNum, fieldName: 'Prefab Build Start Date', message: `Invalid date: ${startStr}` });
        }
      }

      const finishStr = getField('prefab build finish date');
      if (finishStr) {
        const d = tryParseDate(finishStr);
        if (d) {
          update.newFinish = d.toISOString();
        } else {
          errors.push({ rowNumber: rowNum, fieldName: 'Prefab Build Finish Date', message: `Invalid date: ${finishStr}` });
        }
      }

      // Work Days (Reference) → duration in minutes
      const workDaysStr = getField('work days (reference)');
      if (workDaysStr) {
        const v = parseFloat(workDaysStr);
        if (!isNaN(v) && isFinite(v)) {
          update.newDurationMinutes = v * MINUTES_PER_WORK_DAY;
        } else {
          errors.push({
            rowNumber: rowNum,
            fieldName: 'Work Days (Reference)',
            message: `Invalid numeric value: ${workDaysStr}`,
          });
          continue;
        }
      }

      // Required → Deadline
      const requiredStr = getField('required');
      if (requiredStr) {
        const d = tryParseDate(requiredStr);
        if (d) update.newDeadline = d.toISOString();
      }

      // Status → PercentComplete
      const statusStr = getField('status');
      if (statusStr) {
        const pct = resolveStatus(statusStr);
        if (pct !== null) {
          update.newPercentComplete = pct;
        } else {
          errors.push({
            rowNumber: rowNum,
            fieldName: 'Status',
            message: `Unrecognized status: ${statusStr}`,
          });
        }
      }

      // Notes = Description + Notes joined with \n
      const description = getField('description');
      const notes = getField('notes');
      const parts: string[] = [];
      if (description) parts.push(description);
      if (notes) parts.push(notes);
      if (parts.length > 0) {
        update.notesAppend = parts.join('\n');
      }

      updates.push(update);
    }

    return createParseResult(updates, errors);
  }
}
